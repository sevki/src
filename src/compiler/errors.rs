use std::ops::Range;

use crate::lexer::Coord;

pub struct Errors<'a>(Vec<lalrpop_util::ErrorRecovery<Coord, crate::lexer::Token<'a>, &'a str>>);

impl<'a> From<Vec<lalrpop_util::ErrorRecovery<Coord, crate::lexer::Token<'a>, &'a str>>>
    for Errors<'a>
{
    fn from(
        value: Vec<lalrpop_util::ErrorRecovery<Coord, crate::lexer::Token<'a>, &'a str>>,
    ) -> Self {
        Self(value)
    }
}

impl<'a> IntoIterator for Errors<'a> {
    type Item = Range<Coord>;

    type IntoIter = <Vec<std::ops::Range<Coord>> as IntoIterator>::IntoIter;

    fn into_iter(self) -> Self::IntoIter {
        self.0
            .into_iter()
            .map(|error| match error.error {
                lalrpop_util::ParseError::InvalidToken { location } => location..location,
                lalrpop_util::ParseError::UnrecognizedEof {
                    location,
                    expected: _,
                } => location..location,
                lalrpop_util::ParseError::UnrecognizedToken { token, expected: _ } => {
                    token.0..token.2
                }
                lalrpop_util::ParseError::ExtraToken { token } => token.0..token.2,
                lalrpop_util::ParseError::User { error: _ } => todo!(),
            })
            .collect::<Vec<_>>()
            .into_iter()
    }
}

fn handle_errors(
    errors: Vec<lalrpop_util::ErrorRecovery<Coord, crate::lexer::Token, &str>>,
    src: &str,
) -> String {
    let mut pretty = String::new();
    let mut last_end = Coord::default();

    for error in errors {
        match error.error {
            lalrpop_util::ParseError::InvalidToken { location: _ } => todo!(),
            lalrpop_util::ParseError::UnrecognizedEof {
                location: _,
                expected: _,
            } => todo!(),
            lalrpop_util::ParseError::UnrecognizedToken { token, expected } => {
                // find the line and column of the start and end tokens,
                // and print the line with a caret pointing to the error
                let start = token.0;
                let end = token.2;
                let start_line = start.line;
                let end_line = end.line;
                let line = &src[start_line..end_line];
                let start_col = start.col;
                let end_col = end.col;
                // pretty.push_str(&src[last_end..start]);
                pretty.push_str(&format!(
                    "error: unexpected token {:?}, expected one of {:?}\n",
                    token.1, expected
                ));
                pretty.push_str(&line);
                pretty.push_str("\n");
                pretty.push_str(&" ".repeat(start_col));
                pretty.push_str(&"^".repeat(end_col - start_col));
                last_end = end;
            }
            lalrpop_util::ParseError::ExtraToken { token: _ } => todo!(),
            lalrpop_util::ParseError::User { error: _ } => todo!(),
        };
    }
    // pretty.push_str(&src[last_end..]);
    pretty
}
