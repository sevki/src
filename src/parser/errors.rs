use crate::lexer::Coord;

pub fn pretty_errors<'input>(
    src: &'input str,
    errors: Vec<lalrpop_util::ErrorRecovery<Coord, crate::lexer::Token<'_>, &str>>,
) -> String {
    let mut pretty = String::new();

    for error in errors {
        match error.error {
            lalrpop_util::ParseError::InvalidToken { location: _ } => todo!(),
            lalrpop_util::ParseError::UnrecognizedEof { location, expected } => {
                let start = location;
                let end = location;
                let start_line = start.line;
                let end_line = end.line;
                let line = &src[start_line..end_line];
                let start_col = start.col;
                let end_col = end.col;
                pretty.push_str(&format!(
                    "error: unexpected end of file, expected one of {:?}\n",
                    expected
                ));
                pretty.push_str(&line);
                pretty.push_str("\n");
                pretty.push_str(&" ".repeat(start_col));
                pretty.push_str(&"^".repeat(end_col - start_col));
            }
            lalrpop_util::ParseError::UnrecognizedToken { token, expected } => {
                // find the line and column of the start and end tokens,
                // and print the line with a caret pointing to the error
                let start = token.0;
                let end = token.2;
                let start_line = start.line;
                let end_line = end.line;
                let line = &src[start_line..end_line];
                let start_col = start.col;
                let end_col = end.col;
                // pretty.push_str(&src[last_end..start]);
                // expected is double quoted, so we need to remove the quotes
                // which ends up printing like   "\"(\""
                let expected: Vec<&str> = expected.iter().map(|x| &x[1..x.len() - 1]).collect();
                pretty.push_str(&format!(
                    "error: unexpected token {:?}, expected one of {:?}\n",
                    token.1, expected
                ));
                pretty.push_str(&line);
                pretty.push_str("\n");
                pretty.push_str(&" ".repeat(start_col));
                pretty.push_str(&"^".repeat(end_col - start_col));
            }
            lalrpop_util::ParseError::ExtraToken { token: _ } => todo!(),
            lalrpop_util::ParseError::User { error: _ } => todo!(),
        };
    }
    // pretty.push_str(&src[last_end..]);
    pretty
}
