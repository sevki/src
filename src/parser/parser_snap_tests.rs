
use crate::parser::errors::pretty_errors;
use insta::assert_snapshot;


#[cfg(test)]
#[okstd::test]
fn test_empty_parser() {
    let input = " ";
    let mut errors = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    assert!(errors.is_empty());
    assert_snapshot!(format!("{:#?}", t.unwrap()), @r###"
    Module(
        [],
    )
    "###);
}

#[okstd::test]
fn test_fn_call_parser_with_multiple_args_and_strings() {
    let input = "fn some()[] {let a = some_fnExpr(1, \"2\", 3)}";
    let mut errors = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    if !errors.is_empty() {
        panic!("{}", pretty_errors(&input, errors));
    }
    assert_snapshot!(format!("{:#?}", t.unwrap()));
}

#[okstd::test]
fn test_fn_def_parser() {
    let input = r#"fn call(a:b, b:c) [throws, awaits, execs] {
        call(1+1)
        let a = 1
    } when throws {
        raise(1)
    }"#;
    let mut errors = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    assert!(errors.is_empty());
    assert_snapshot!(format!("{:#?}", t.unwrap()));
}

#[okstd::test]
fn test_effect() {
    let input = r#"effect VM: async + throws + execs {
        catch() []
        await<T>(f: Future<T>) [] -> T
        exec(arg0: string, args: stringvec) []
    }"#;
    let mut errors = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    assert!(errors.is_empty());
    assert_snapshot!(format!("{:#?}", t.unwrap()));
}

#[okstd::test]
fn test_struct_parser() {
    let input = r#"struct VM {
        a: string
        b: string
    }"#;
    let mut errors = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    assert!(errors.is_empty());
    assert_snapshot!(format!("{:#?}", t.unwrap()));
}

#[okstd::test]
fn test_enum_parser() {
    let input = r#"use { exec } from host

    effect Make: async + throws + execs + reads + writes {
        catch() [throws]
        await<T>(f: Future<T>) [async, throws] -> T
        exec(arg0: string, args: stringvec) [Make] -> i32
        read(name: string) [reads] -> string
        write(name: string, value: string) [writes]
    }
    
    struct Local {
        host: host
    }
    
    impl Make for Local {
        fn catch(self) [throws] {
        } 
        fn await<T>(f: Future<T>) [async, trhows] -> T {
            yield()
        }
        fn exec(self, arg0: string, args: vec<string>) [Vm] -> i32 {
            self.host.read("jobserver")
            if self.host.exec(arg0, args) {
                raise(1)
            }
        }
    }"#;
    let mut errors: Vec<lalrpop_util::ErrorRecovery<usize, crate::lexer::Token, &str>> = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    if !errors.is_empty() {
        panic!("{}", pretty_errors(&input, errors));
    }
    assert_snapshot!(format!("{:#?}", t.unwrap()));
}
