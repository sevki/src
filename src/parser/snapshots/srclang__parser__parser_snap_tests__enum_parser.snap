---
source: src/parser/parser_snap_tests.rs
expression: "format!(\"{:#?}\", t.unwrap())"
---
Module(
    [
        UseDef(
            UseDef(
                [
                    Ident(
                        "exec",
                        None,
                    ),
                ],
                Ident(
                    "host",
                    None,
                ),
            ),
        ),
        EffectDef(
            EffectDef(
                Ident(
                    "Make",
                    None,
                ),
                [
                    Ident(
                        "async",
                        None,
                    ),
                    Ident(
                        "throws",
                        None,
                    ),
                    Ident(
                        "execs",
                        None,
                    ),
                    Ident(
                        "reads",
                        None,
                    ),
                    Ident(
                        "writes",
                        None,
                    ),
                ],
                Block(
                    [
                        Prototype {
                            name: Ident(
                                "catch",
                                None,
                            ),
                            args: [],
                            ret: None,
                            effects: [
                                Ident(
                                    "throws",
                                    None,
                                ),
                            ],
                        },
                        Prototype {
                            name: Ident(
                                "await",
                                Some(
                                    [
                                        Ident(
                                            "T",
                                            None,
                                        ),
                                    ],
                                ),
                            ),
                            args: [
                                Field(
                                    Field(
                                        Ident(
                                            "f",
                                            None,
                                        ),
                                        Ident(
                                            "Future",
                                            Some(
                                                [
                                                    Ident(
                                                        "T",
                                                        None,
                                                    ),
                                                ],
                                            ),
                                        ),
                                    ),
                                ),
                            ],
                            ret: Some(
                                Ident(
                                    "T",
                                    None,
                                ),
                            ),
                            effects: [
                                Ident(
                                    "async",
                                    None,
                                ),
                                Ident(
                                    "throws",
                                    None,
                                ),
                            ],
                        },
                        Prototype {
                            name: Ident(
                                "exec",
                                None,
                            ),
                            args: [
                                Field(
                                    Field(
                                        Ident(
                                            "arg0",
                                            None,
                                        ),
                                        Ident(
                                            "string",
                                            None,
                                        ),
                                    ),
                                ),
                                Field(
                                    Field(
                                        Ident(
                                            "args",
                                            None,
                                        ),
                                        Ident(
                                            "stringvec",
                                            None,
                                        ),
                                    ),
                                ),
                            ],
                            ret: Some(
                                Ident(
                                    "i32",
                                    None,
                                ),
                            ),
                            effects: [
                                Ident(
                                    "Make",
                                    None,
                                ),
                            ],
                        },
                        Prototype {
                            name: Ident(
                                "read",
                                None,
                            ),
                            args: [
                                Field(
                                    Field(
                                        Ident(
                                            "name",
                                            None,
                                        ),
                                        Ident(
                                            "string",
                                            None,
                                        ),
                                    ),
                                ),
                            ],
                            ret: Some(
                                Ident(
                                    "string",
                                    None,
                                ),
                            ),
                            effects: [
                                Ident(
                                    "reads",
                                    None,
                                ),
                            ],
                        },
                        Prototype {
                            name: Ident(
                                "write",
                                None,
                            ),
                            args: [
                                Field(
                                    Field(
                                        Ident(
                                            "name",
                                            None,
                                        ),
                                        Ident(
                                            "string",
                                            None,
                                        ),
                                    ),
                                ),
                                Field(
                                    Field(
                                        Ident(
                                            "value",
                                            None,
                                        ),
                                        Ident(
                                            "string",
                                            None,
                                        ),
                                    ),
                                ),
                            ],
                            ret: None,
                            effects: [
                                Ident(
                                    "writes",
                                    None,
                                ),
                            ],
                        },
                    ],
                ),
            ),
        ),
        StructDef(
            StructDef(
                Ident(
                    "Local",
                    None,
                ),
                Block(
                    [
                        Field(
                            Ident(
                                "host",
                                None,
                            ),
                            Ident(
                                "host",
                                None,
                            ),
                        ),
                    ],
                ),
            ),
        ),
        ImplDef(
            ImplDef(
                Ident(
                    "Make",
                    None,
                ),
                Some(
                    Ident(
                        "Local",
                        None,
                    ),
                ),
                Block(
                    [
                        FnDef(
                            FnDef(
                                Prototype {
                                    name: Ident(
                                        "catch",
                                        None,
                                    ),
                                    args: [
                                        Reciever,
                                    ],
                                    ret: None,
                                    effects: [
                                        Ident(
                                            "throws",
                                            None,
                                        ),
                                    ],
                                },
                                Block(
                                    [],
                                ),
                                [],
                            ),
                        ),
                        FnDef(
                            FnDef(
                                Prototype {
                                    name: Ident(
                                        "await",
                                        Some(
                                            [
                                                Ident(
                                                    "T",
                                                    None,
                                                ),
                                            ],
                                        ),
                                    ),
                                    args: [
                                        Field(
                                            Field(
                                                Ident(
                                                    "f",
                                                    None,
                                                ),
                                                Ident(
                                                    "Future",
                                                    Some(
                                                        [
                                                            Ident(
                                                                "T",
                                                                None,
                                                            ),
                                                        ],
                                                    ),
                                                ),
                                            ),
                                        ),
                                    ],
                                    ret: Some(
                                        Ident(
                                            "T",
                                            None,
                                        ),
                                    ),
                                    effects: [
                                        Ident(
                                            "async",
                                            None,
                                        ),
                                        Ident(
                                            "trhows",
                                            None,
                                        ),
                                    ],
                                },
                                Block(
                                    [
                                        FnCall(
                                            FnCall(
                                                Ident(
                                                    "yield",
                                                    None,
                                                ),
                                                [],
                                            ),
                                        ),
                                    ],
                                ),
                                [],
                            ),
                        ),
                        FnDef(
                            FnDef(
                                Prototype {
                                    name: Ident(
                                        "exec",
                                        None,
                                    ),
                                    args: [
                                        Reciever,
                                        Field(
                                            Field(
                                                Ident(
                                                    "arg0",
                                                    None,
                                                ),
                                                Ident(
                                                    "string",
                                                    None,
                                                ),
                                            ),
                                        ),
                                        Field(
                                            Field(
                                                Ident(
                                                    "args",
                                                    None,
                                                ),
                                                Ident(
                                                    "vec",
                                                    Some(
                                                        [
                                                            Ident(
                                                                "string",
                                                                None,
                                                            ),
                                                        ],
                                                    ),
                                                ),
                                            ),
                                        ),
                                    ],
                                    ret: Some(
                                        Ident(
                                            "i32",
                                            None,
                                        ),
                                    ),
                                    effects: [
                                        Ident(
                                            "Vm",
                                            None,
                                        ),
                                    ],
                                },
                                Block(
                                    [
                                        FieldAccess(
                                            FieldAccess(
                                                FieldAccess(
                                                    FieldAccess(
                                                        FieldAccess(
                                                            FieldAccess(
                                                                Ident(
                                                                    Ident(
                                                                        "self",
                                                                        None,
                                                                    ),
                                                                ),
                                                                Ident(
                                                                    Ident(
                                                                        "host",
                                                                        None,
                                                                    ),
                                                                ),
                                                            ),
                                                        ),
                                                        FnCall(
                                                            FnCall(
                                                                Ident(
                                                                    "read",
                                                                    None,
                                                                ),
                                                                [
                                                                    String(
                                                                        "jobserver",
                                                                    ),
                                                                ],
                                                            ),
                                                        ),
                                                    ),
                                                ),
                                                Ident(
                                                    Ident(
                                                        "await",
                                                        None,
                                                    ),
                                                ),
                                            ),
                                        ),
                                        Branch(
                                            Branch(
                                                FieldAccess(
                                                    FieldAccess(
                                                        FieldAccess(
                                                            FieldAccess(
                                                                FieldAccess(
                                                                    FieldAccess(
                                                                        Ident(
                                                                            Ident(
                                                                                "self",
                                                                                None,
                                                                            ),
                                                                        ),
                                                                        Ident(
                                                                            Ident(
                                                                                "host",
                                                                                None,
                                                                            ),
                                                                        ),
                                                                    ),
                                                                ),
                                                                FnCall(
                                                                    FnCall(
                                                                        Ident(
                                                                            "exec",
                                                                            None,
                                                                        ),
                                                                        [
                                                                            Ident(
                                                                                Ident(
                                                                                    "arg0",
                                                                                    None,
                                                                                ),
                                                                            ),
                                                                            Ident(
                                                                                Ident(
                                                                                    "args",
                                                                                    None,
                                                                                ),
                                                                            ),
                                                                        ],
                                                                    ),
                                                                ),
                                                            ),
                                                        ),
                                                        Ident(
                                                            Ident(
                                                                "await",
                                                                None,
                                                            ),
                                                        ),
                                                    ),
                                                ),
                                                [
                                                    (
                                                        Bool(
                                                            true,
                                                        ),
                                                        Block(
                                                            [
                                                                FnCall(
                                                                    FnCall(
                                                                        Ident(
                                                                            "raise",
                                                                            None,
                                                                        ),
                                                                        [
                                                                            Integer(
                                                                                1,
                                                                            ),
                                                                        ],
                                                                    ),
                                                                ),
                                                            ],
                                                        ),
                                                    ),
                                                ],
                                            ),
                                        ),
                                    ],
                                ),
                                [],
                            ),
                        ),
                    ],
                ),
            ),
        ),
    ],
)
